src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -845,29 +845,32 @@ private int ReadBuffer(Span<char> userBuffer, out bool readToUserBuffer)
             return sb.ToString();
         }
 
-        public override Task<string?> ReadLineAsync()
+        public override Task<string?> ReadLineAsync() =>
+            ReadLineAsync(default).AsTask();
+
+        public override ValueTask<string?> ReadLineAsync(CancellationToken cancellationToken)
         {
             // If we have been inherited into a subclass, the following implementation could be incorrect
             // since it does not call through to Read() which a subclass might have overridden.
             // To be safe we will only use this implementation in cases where we know it is safe to do so,
             // and delegate to our base class (which will call into Read) when we are not sure.
             if (GetType() != typeof(StreamReader))
             {
-                return base.ReadLineAsync();
+                return base.ReadLineAsync(cancellationToken);
             }
 
             ThrowIfDisposed();
             CheckAsyncTaskInProgress();
 
-            Task<string?> task = ReadLineAsyncInternal();
+            Task<string?> task = ReadLineAsyncInternal(cancellationToken);
             _asyncReadTask = task;
 
-            return task;
+            return new ValueTask<string?>(task);
         }
 
-        private async Task<string?> ReadLineAsyncInternal()
+        private async Task<string?> ReadLineAsyncInternal(CancellationToken cancellationToken)
```

Comment:
I agree, but it seemed like allocating a `Task` (to store in `_asyncReadTask`) was unavoidable. However, I see now that the only use of `_asyncReadTask` is checking `.IsCompleted`, so it should be safe to switch that to a `ValueTask`.

Reply:
Except that `ValueTask<string?>` can't be stored in a `ValueTask` field, so I think we're back to the original code.

Reply:
> Except that ValueTask<string?> can't be stored in a ValueTask field

Did you mean a `Task` field? That is correct. 

Perhaps we should do what other methods are doing:

`Task<int> ReadAsync(char[] buffer, int index, int count)` converts a VT into Task and stores it in the field:

https://github.com/dotnet/runtime/blob/4a099dc0eea7acd5d9a8d4c62cb89f24da6bf7ff/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs#L991-L992

`ValueTask<int> ReadAsync(Memory<char> buffer, CancellationToken cancellationToken = default)` does check if
there is any async **task** in progress:

https://github.com/dotnet/runtime/blob/4a099dc0eea7acd5d9a8d4c62cb89f24da6bf7ff/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs#L1006

but does not modify the value of the `_asyncReadTask` field.

@stephentoub I assume that this is correct for `ValueTask`-returning methods?

Reply:
> Did you mean a Task field? That is correct.

I was thinking out loud about converting the `_asyncReadTask` field to be `ValueTask`, in order to avoid the `Task` allocation (which would happen in `.AsTask()`) entirely (understanding that to be the original goal). But doing that conversion wouldn't work, since `ValueTask<T>` can't be assigned to a `ValueTask` field. If I keep `_asyncReadTask` as `Task`, then a `Task` allocation has to happen (via `AsTask()`) for any `ValueTask`-returning method whose result is stored in that field, possibly negating the benefits of making that method return `ValueTask` in the first place.

> Perhaps we should do what other methods are doing: ... converts a VT into Task and stores it in the field:

Yes, can do.

Reply:
> does check if there is any async task in progress but does not modify the value of the _asyncReadTask field
> @stephentoub I assume that this is correct for ValueTask-returning methods?

Honestly that seems like a bug.  The purpose of that field is to help a user detect misuse, erroneously invoking an async operation on the reader without having awaited the previous operation.  If we want such detection to remain, it should really remain for any async operation, regardless of whether it returns `Task` or `ValueTask`.  We could also say it's not worth the expense to detect such misuse, get rid of the detection entirely, and if the developer misuses it, that's on them.